Balder/looking for missing join

baldersheim opened the pull request:
@vekterli PR
Only the the 3rd commit contains sematics.
I also wonder about join. Now it returns false if stop has not been called and succeeded.
I suggest it either calls stop or throws an exception.

baldersheim committed:
Cleanup while looking for missing join.

baldersheim committed:
No longer set unused parameter

baldersheim committed:
Avoid that we miss a stop that happens while we are STARTING.

baldersheim committed:
Put the filereferences in ''

baldersheim committed:
Earlier conclusion about issue with too ealy stop was incorrect. However adding sanity checks to catch incorrect usage.

baldersheim committed:
You can only start threads that operates on this in the constructor if you are final.

baldersheim committed:
Remove unused method declaration and allign code.

vekterli commented on staging_vespalib/src/vespa/vespalib/util/document_runnable.cpp:
```diff
@@ -2,15 +2,19 @@
       _state(NOT_RUNNING)
 {
-    if (pool) start(*pool);
+    // if (pool) start(*pool);
```
Leftover commented out code

baldersheim replied:
Done

vekterli commented on staging_vespalib/src/vespa/vespalib/util/document_runnable.cpp:
```diff
@@ -42,7 +46,7 @@ bool Runnable::onStop()
 {
     vespalib::MonitorGuard monitor(_stateLock);
-    if (_state == STARTING || _state == RUNNING) return false;
+    assert ((_state != STARTING) && (_state != RUNNING));
```
Would it would make sense to convert the similar return branch in `stop()` to an assertion as well? These kinds of silent returns are perfect for hiding wrongful use of the class API.

Though I fear that we have usages where we conditionally start something on init, but always stop on destruction, meaning that an assertion might core the process during an exception unwind on startup...

baldersheim replied:
That might be a next step.

baldersheim commented:
The takeaway here is that I have added assert that stop/join has been called and that state is correct on destruction. I also removed possibility to start a Runnable in the constructor as that is simply wrong.

vekterli reviewed (approved):
:+1: only minor comments/questions

baldersheim committed:
Remove leftover code.
